ci-operator/config/openshift/os/openshift-os-master.yaml

```diff
@@ -139,9 +139,12 @@ resources:
       memory: 3Gi
 tests:
 - as: validate
-  commands: ./ci/validate.sh
+  commands: |
+    #!/bin/bash
+    set -xeuo
+    ./ci/validate.sh
   container:
-    from: src
+    from: coreos_coreos-assembler_latest
```

Comment:
Changing this to `coreos_coreos-assembler_latest` would probably break because `src` points to what's defined under `.images.build_root`.

Reply:
Now I'm a little confused as well! But I think I have an understanding of what's going on here and explain the source of my confusion (and hopefully make you less confused as well!):

- The Dockerfile for `openshift/os` is in `ci/Dockerfile` and is what becomes `build-test-qemu-img`. We probably should make this the `.build_root` part of the CI config, but that's a separate concern and not relevant right now. The resulting image is what the COSA build scripts use to run. It has the scripts and the layering test binary from `openshift/os` layered on top of the `coreos-assembler` image. This is why I thought we should use that image to run `./ci/validate.sh`.
- The `src` image is `registry.ci.openshift.org/coreos/fcos-buildroot:testing-devel`, produced by the [`coreos/fedora-coreos-config` repo](https://github.com/coreos/fedora-coreos-config). Interestingly, the `validate` step is the only part of the CI config that uses this image. None of the other image builds or tests directly consume or use this image.
- The "magic" is that before the `validate` test runs, a bunch of init containers are run. Amongst those is a `cloneRefs` step (`$ curl -s 'https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/28790/rehearse-28790-pull-ci-openshift-os-master-validate/1529165858037829632/artifacts/ci-operator-step-graph.json' | jq -r '.[1].manifests[0].spec.initContainers[].name'`) which clones `openshift/os` (and the `fedora-coreos-config` submodule).

So in a nutshell, what's happening in the `validate` step is we're taking `registry.ci.openshift.org/coreos/coreos-assembler:latest`, cloning `openshift/os` (and the `fedora-coreos-config` submodule) into it and running `./ci/validate.sh`. My confusion came from forgetting about the `cloneRefs` step and thinking that since `coreos-assembler:latest` doesn't have the `./ci/validate.sh` script present that it would fail.

Reply:
Thanks for the detailed information, Zack!  It certainly helps improve my understanding of the flow of the jobs.

>  Interestingly, the validate step is the only part of the openshift/os CI config that uses this image. None of the other image builds or tests directly consume or use this image.

Should we drop the use of the `fcos-buildroot` as part of this PR?  Would we need to reconfigure `build_root` to point to another image?

Is the `cloneRefs` step logged in any of the test artifacts?

If you think the PR is good to go as is, please drop an `/approve` if you can.

Reply:
> Should we drop the use of the fcos-buildroot as part of this PR? Would we need to reconfigure build_root to point to another image?

We could configure it to point to `ci/Dockerfile` and build that. We'd then need to replace `build-test-qemu-img` with `src`. There's no rush in doing that, so if we want to do that, let's handle that as a separate PR.

> Is the cloneRefs step logged in any of the test artifacts?

Unfortunately it's not. From what I can tell, the container used purposely does not create logs (although I don't know why). You _might_ be able to catch it while the job is running if you click through to the console for the CI namespace. The only place I was really able to find it was in the [job test steps artifact](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/28790/rehearse-28790-pull-ci-openshift-os-master-validate/1529165858037829632/artifacts/ci-operator-step-graph.json) and even then, I had to know that it was an init container.

> If you think the PR is good to go as is, please drop an /approve if you can.

Will do.